lib/markdown-preview-view.js

```diff
@@ -0,0 +1,532 @@
+    return Array.prototype.slice
+      .apply(textEditorStyles.childNodes)
+      .map(styleElement => styleElement.innerText)
+  }
+
+  getMarkdownPreviewCSS () {
+    const markdownPreviewRules = []
+    const ruleRegExp = /\.markdown-preview/
+    const cssUrlRegExp = /url\(atom:\/\/markdown-preview\/assets\/(.*)\)/
+
+    for (const stylesheet of this.getDocumentStyleSheets()) {
+      if (stylesheet.rules != null) {
+        for (const rule of stylesheet.rules) {
+          // We only need `.markdown-review` css
+          if (
+            (rule.selectorText != null
+              ? rule.selectorText.match(ruleRegExp)
+              : undefined) != null
+          ) {
+            markdownPreviewRules.push(rule.cssText)
+          }
+        }
+      }
+    }
+
+    return markdownPreviewRules
+      .concat(this.getTextEditorStyles())
+      .join('\n')
+      .replace(/atom-text-editor/g, 'pre.editor-colors')
+      .replace(/:host/g, '.host') // Remove shadow-dom :host selector causing problem on FF
+      .replace(cssUrlRegExp, function (match, assetsName, offset, string) {
+        // base64 encode assets
+        const assetPath = path.join(__dirname, '../assets', assetsName)
+        const originalData = fs.readFileSync(assetPath, 'binary')
+        const base64Data = Buffer.from(originalData, 'binary').toString('base64')
+        return `url('data:image/jpeg;base64,${base64Data}')`
+      })
+  }
+
+  showError (result) {
+    let failureMessage
+    this.element.textContent = ''
+    const h2 = document.createElement('h2')
+    h2.textContent = 'Previewing Markdown Failed'
+    this.element.appendChild(h2)
+    if ((failureMessage = result != null ? result.message : undefined)) {
+      const h3 = document.createElement('h3')
+      h3.textContent = failureMessage
+      return this.element.appendChild(h3)
+    }
+  }
+
+  showLoading () {
+    this.loading = true
+    this.element.textContent = ''
+    const div = document.createElement('div')
+    div.classList.add('markdown-spinner')
+    div.textContent = 'Loading Markdown\u2026'
+    return this.element.appendChild(div)
+  }
+
+  selectAll () {
+    if (this.loading) {
+      return
+    }
+
+    const selection = window.getSelection()
+    selection.removeAllRanges()
+    const range = document.createRange()
+    range.selectNodeContents(this.element)
+    return selection.addRange(range)
+  }
+
+  copyToClipboard () {
+    if (this.loading) {
+      return
+    }
+
+    const selection = window.getSelection()
+    const selectedText = selection.toString()
+    const selectedNode = selection.baseNode
+
+    // Use default copy event handler if there is selected text inside this view
+    if (
+      selectedText &&
+      selectedNode != null &&
+      (this.element === selectedNode || this.element.contains(selectedNode))
+    ) {
+      return atom.clipboard.write(selectedText)
+    } else {
+      return this.getHTML(function (error, html) {
```

Comment:
:fire: return

Reply:
This one is needed, since `getHTML` returns a `Promise` that needs to be awaited (I spent too much time debugging a failing test caused by this one 😅).